crates/flox-rust-sdk/src/models/search.rs

```diff
@@ -0,0 +1,358 @@
+/// without an enclosing `[]`, so the results returned by `pkgdb` can't be
+/// directly deserialized to a JSON object. To parse the results you should
+/// use the provided `TryFrom` impl.
+#[derive(Debug, Clone, Serialize)]
+pub struct SearchResults {
+    pub results: Vec<SearchResult>,
+}
+
+impl TryFrom<&[u8]> for SearchResults {
+    type Error = SearchError;
+
+    // Note, this impl isn't actually used in the CLI, it's leftover from a previous iteration on the design.
+    // It still works, so we should keep it around. It may prove useful for testing or something.
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let mut results = Vec::new();
+        for maybe_line in bytes.lines() {
+            let text = maybe_line.map_err(SearchError::ParseStdout)?;
+            match serde_json::from_str(&text) {
+                Ok(search_result) => results.push(search_result),
+                Err(_) => {
+                    // TODO: Errors are currently emitted to stdout as JSON, but there's no spec for the errors.
+                    //       For now if we can't turn the text into a SearchResult, we assume that it's an
+                    //       error message. If parsing that into a serde_json::Value fails, something else went
+                    //       pretty wrong.
+                    //
+                    //       Once there's a spec for the error messages we can parse this into a typed container.
+                    let err = Value::from_str(&text)?;
+                    return Err(SearchError::PkgDb(err));
+                },
+            };
+        }
+        Ok(SearchResults { results })
+    }
+}
+
+/// Read search results from a buffered input source.
+///
+/// Fails fast on reading the first error.
+pub fn collect_results(
+    result_reader: impl BufRead,
+    _err_reader: impl BufRead,
+) -> Result<SearchResults, SearchError> {
+    let mut results = Vec::new();
+    for maybe_line in result_reader.lines() {
+        let text = maybe_line.map_err(SearchError::ParseStdout)?;
+        match serde_json::from_str(&text) {
+            Ok(search_result) => results.push(search_result),
+            Err(_) => {
+                // TODO: Errors are currently emitted to stdout as JSON, but there's no spec for the errors.
+                //       For now if we can't turn the text into a SearchResult, we assume that it's an
+                //       error message. If parsing that into a serde_json::Value fails, something else went
+                //       pretty wrong.
+                //
+                //       Once there's a spec for the error messages we can parse this into a typed container.
+                let err = Value::from_str(&text)?;
+                return Err(SearchError::PkgDb(err));
+            },
+        };
+    }
+    Ok(SearchResults { results })
+}
+
+/// Calls `pkgdb` to get search results
+pub fn do_search(search_params: &str) -> Result<(SearchResults, ExitStatus), SearchError> {
```

Comment:
suggestion (non-blocking): accept `SearchParams` rather than `&str`

Reply:
This method is intended to be just a wrapper for the I/O, I consider it out of scope for this function to handle an error converting the `SearchParams` to a string.

Reply:
Not a strong opinion though

Reply:
Fixed